src/main.cpp

```diff
@@ -816,14 +823,58 @@ static bool run_dedicated_server(const GameParams &game_params, const Settings &
 	if (cmd_args.exists("migrate"))
 		return migrate_database(game_params, cmd_args);
 
-	// Create server
-	Server server(game_params.world_path, game_params.game_spec, false,
-		bind_addr.isIPv6());
-	server.start(bind_addr);
+	if (cmd_args.exists("terminal")) {
+#if USE_CURSES
+		bool name_ok = true;
+		std::string admin_nick = g_settings->get("name");
 
-	// Run server
-	bool &kill = *porting::signal_handler_killstatus();
-	dedicated_server_loop(server, kill);
+		name_ok = name_ok && !admin_nick.empty();
+		name_ok = name_ok && string_allowed(admin_nick, PLAYERNAME_ALLOWED_CHARS);
+
+		if (!name_ok) {
+			errorstream << "No valid name given for admin. "
+				<< "Please check your minetest.conf that it "
+				<< "contains a 'name = ' to your main admin account."
+				<< std::endl;
+			return false;
+		}
+		ChatInterface iface;
+		bool &kill = *porting::signal_handler_killstatus();
+
+		// Create server
+		Server server(game_params.world_path,
+			game_params.game_spec, false, bind_addr.isIPv6(), &iface);
+
+		g_term_console.setup(&iface, &kill, admin_nick);
+
+		g_term_console.start();
+
+		server.start(bind_addr);
+		// Run server
+		dedicated_server_loop(server, kill);
+
+		// Tell the console to stop, and wait for it to finish,
+		// only then leave context and free iface
+		g_term_console.stop();
+		g_term_console.wait();
+
+		g_term_console.clearKillStatus();
+	} else {
+#else
```

Comment:
I really do not like how this preprocessor directive cuts into the middle of a syntactical structure...